src/Discord/Internal/Types/ApplicationCommands.hs

```diff
@@ -0,0 +1,396 @@
+    object
+      [ (name, value)
+        | (name, Just value) <-
+            [ ("type", toMaybeJSON optionType),
+              ("name", toMaybeJSON optionName),
+              ("description", toMaybeJSON optionDescription),
+              ("required", toJSON <$> optionRequired),
+              ("choices", toJSON <$> optionChoices),
+              ("options", toJSON <$> optionOptions),
+              ("channel_types", toJSON <$> optionChannelTypes),
+              ("min_val", toJSON <$> optionMinVal),
+              ("max_val", toJSON <$> optionMaxVal),
+              ("autocomplete", toJSON <$> optionAutocomplete)
+            ]
+      ]
+
+instance FromJSON ApplicationCommandOption where
+  parseJSON =
+    withObject
+      "ApplicationCommandOption"
+      ( \v ->
+          ApplicationCommandOption
+            <$> v .: "type"
+            <*> v .: "name"
+            <*> v .: "description"
+            <*> v .:? "required"
+            <*> v .:? "choices"
+            <*> v .:? "options"
+            <*> v .:? "channel_types"
+            <*> v .:? "min_val"
+            <*> v .:? "max_val"
+            <*> v .:? "autocomplete"
+      )
+
+-- | What type of command option. Can represent a wide variety of types, so
+-- please check out the documentation below.
+--
+-- https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-type
+data ApplicationCommandOptionType
+  = -- | A subcommand. It can take further options, excluding sub commands and
+    -- sub command groups.
+    SUB_COMMAND
+  | -- | A subcommand group. It can take further options, excluding sub command
+    -- groups.
+    SUB_COMMAND_GROUP
+  | -- | Can typically be provided with default values.
+    STRING
+  | -- | Can typically be provided with default values, and possibly with
+    -- minimum and maximum values.
+    INTEGER
+  | BOOLEAN
+  | USER
+  | -- | Can be limited in the types of the channel allowed.
+    CHANNEL
+  | ROLE
+  | -- | Users and roles.
+    MENTIONABLE
+  | -- | Can typically be provided with default values, and possibly with
+    -- minimum and maximum values. Represents a double.
+    NUMBER
+  deriving (Show, Read, Data, Eq)
+
+instance Enum ApplicationCommandOptionType where
+  fromEnum SUB_COMMAND = 1
+  fromEnum SUB_COMMAND_GROUP = 2
+  fromEnum STRING = 3
+  fromEnum INTEGER = 4
+  fromEnum BOOLEAN = 5
+  fromEnum USER = 6
+  fromEnum CHANNEL = 7
+  fromEnum ROLE = 8
+  fromEnum MENTIONABLE = 9
+  fromEnum NUMBER = 10
+  toEnum a = fromJust $ lookup a table
+    where
+      table = makeTable SUB_COMMAND
+
+instance ToJSON ApplicationCommandOptionType where
+  toJSON = toJSON . fromEnum
+
+instance FromJSON ApplicationCommandOptionType where
+  parseJSON = withScientific "ApplicationCommandOptionType" (return . toEnum . round)
+
+-- | Utility data type to store integers or doubles.
+data IntDouble = IDI Int | IDD Double
+  deriving (Show, Read, Eq)
```

Comment:
Is this utility necessary? A 64bit Double can represent ~52 bit integers 

https://stackoverflow.com/questions/759201/representing-integers-in-doubles

Reply:
I guess the question is do we want to spit out 64bit doubles all the time or allow some things to be integers or doubles?

Reply:
Good point. Should probably allow `int`, `double`, `string` different options. This is find for the backend, but maybe the user frontend could abstract using a nicer ADT? 

`ApplicationCommandOption` has some redundancies, like what happens if the type is set to `string`, but `int` choices are given? And `optionMinVal` for string. 

Fine for the backend, (match the json structure), but ideally the frontend is nicer eventually.

Reply:
I've come up with the solution (which Aeson already had funnily enough)

We just use the Scientific type, and then people can convert out of that if they wish.

When we get around to making a nice front end for all this stuff, we can lock in the choices, but I think this is the best approach for now - honestly "not caring" about the type being given to the JSON, since JSON doesn't differentiate between number and integer anyway